pkg/utils/tsoutil/tso_proto_factory.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsoutil
+
+import (
+	"context"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/tsopb"
+	"google.golang.org/grpc"
+)
+
+// ProtoFactory is the abstract factory for creating tso related data structures defined in the gPRC service
+type ProtoFactory interface {
+	createForwardStream(ctx context.Context, client *grpc.ClientConn) (tsoStream, context.CancelFunc, error)
+}
+
+// TSOProtoFactory is the abstract factory for creating tso related data structures defined in the TSO gPRC service
+type TSOProtoFactory struct {
+}
+
+// PDProtoFactory is the abstract factory for creating tso related data structures defined in the PD gPRC service
+type PDProtoFactory struct {
+}
+
+func (s *TSOProtoFactory) createForwardStream(ctx context.Context, clientConn *grpc.ClientConn) (tsoStream, context.CancelFunc, error) {
+	done := make(chan struct{})
+	cctx, cancel := context.WithCancel(ctx)
+	go checkStream(cctx, cancel, done)
+	forwardStream, err := tsopb.NewTSOClient(clientConn).Tso(cctx)
+	done <- struct{}{}
+	return &tsoTSOStream{forwardStream}, cancel, err
+}
+
+func (s *PDProtoFactory) createForwardStream(ctx context.Context, clientConn *grpc.ClientConn) (tsoStream, context.CancelFunc, error) {
+	done := make(chan struct{})
+	cctx, cancel := context.WithCancel(ctx)
+	go checkStream(cctx, cancel, done)
+	forwardStream, err := pdpb.NewPDClient(clientConn).Tso(cctx)
+	done <- struct{}{}
+	return &pdTSOStream{forwardStream}, cancel, err
+}
+
+type tsoStream interface {
+	// process sends a request and receives the response through the stream
+	process(clusterID uint64, count, keyspaceID, keyspaceGroupID uint32, dcLocation string) (tsoResponse, error)
+}
+
+type tsoTSOStream struct {
```

Comment:
how about just tsoStream or some better naming?

Reply:
You are right. Because it's in tsoutil package, we can actually trim the tso prefix.
Renamed:
tsoStream -> stream
tsoTSOStream -> tsoStream
pdTSOStream -> pdStream